lib/puppet/util/windows/adsi.rb

```diff
@@ -94,6 +94,21 @@ def execquery(query)
       wmi_connection.execquery(query)
     end
 
+    # https://docs.microsoft.com/en-us/dotnet/api/microsoft.powershell.commands.domainrole?view=powershellsdk-1.1.0
+    DOMAIN_ROLES = {
+      STANDALONE_WORKSTATION: 0,
+      MEMBER_WORKSTATION: 1,
+      STANDALONE_SERVER: 2,
+      MEMBER_SERVER: 3,
+      BACKUP_DOMAIN_CONTROLLER: 4,
+      PRIMARY_DOMAIN_CONTROLLER: 5,
+    }
+
+    def domain_role
```

Comment:
This method should memoize the result. Calling WMI is expensive and the DomainRole of the current computer will not change during the course a Puppet run.